lib/detectors/FlashLoan/FlashloanDetector.h

```diff
@@ -0,0 +1,126 @@
+//
+// Created by Nicholas Brown on 1/19/23.
+//
+
+#ifndef VANGUARD_FLASHLOANDETECTOR_H
+#define VANGUARD_FLASHLOANDETECTOR_H
+
+#include <analysis/taint-analysis/TaintAnalysis.h>
+#include <domain/libBlockchain/BlockchainModel.h>
+#include <detectors/UniverseDetector.h>
+#include <iostream>
+#include <unordered_map>
+
+
+namespace vanguard {
+    template <typename Domain>
+    class FlashloanDetector : public UniverseDetector<Domain> {
+    public:
+        using Function = typename Domain::Function;
+        using Observer = typename TaintAnalysis::Observer;
+
+        struct BalanceTransferResolver : public InstructionClassVisitor<Domain> {
+            void visit(const CallIns<Domain> &v) override {
+                if(v.name().find("fun_balanceOf_") != std::string::npos) {
+                    balanceOfReturns.push_back(v.result()->unwrap());
+                }
+                if(v.name().find("fun_transferFrom_") != std::string::npos ||
+                    v.name().find("fun_transfer_") != std::string::npos) {
+                    const llvm::Value *lastArg;
+                    for (int i = 0; i < v.numOperands() - 1 ; i++) {
+                        lastArg = v.operandAt(i)->unwrap();
+
+                    }
+                    transferAmts.push_back(lastArg);
+                }
+            }
+
+            std::vector<const llvm::Value *> balanceOfReturns;
+            std::vector<const llvm::Value *> transferAmts;
+        };
+
+        class BalanceTransferObserver : public Observer {
+        public:
+            BalanceTransferObserver(std::unordered_set<string> functions) : contractFns(functions) {}
+            void onUnreachable(SVF::ProgSlice *slice) override {}
+            void onPartiallyUnreachable(SVF::ProgSlice *slice) override {}
+            void onReachable(SVF::ProgSlice *slice) override {
+                for(TaintAnalysis::SVFGNodeSetIter it = slice->forwardSliceBegin(), eit = slice->forwardSliceEnd(); it != eit; it++) {
+                    std::string nodeFnName = (*it)->getFun()->getName();
+                    for (auto fnName : contractFns) {
+                        if (nodeFnName.find(fnName) != std::string::npos) {
+                            potentiallyVulnerable.insert(fnName);
+                        }
+                    }
+                }
+                vulnerable = true;
+            }
+            bool isVulnerable() {
+                return vulnerable;
+            }
+            std::unordered_set<std::string> getPotentiallyVulnerable() {
```

Comment:
nit: to allow the caller to take by reference, I would change this to:
```suggestion
            const std::unordered_set<std::string> &getPotentiallyVulnerable() const {
```